procs/riscy-lib/RVUart.bsv

```diff
@@ -0,0 +1,136 @@
+import ClientServer::*;
+import RS232::*;
+import Vector::*;
+import GetPut::*;
+import Vector::*;
+import BuildVector::*;
+import Ehr::*;
+
+import Abstraction::*;
+
+interface RVUart#(numeric type cores);
+  // Internal connections
+  interface UncachedMemServer memifc;
+
+  // external connections for rx/tx (to be connected to a UART
+  (* prefix = "" *)
+  interface RS232 uart_pins;
+  //method Action rx (Bit#(1) x);
+  // method Bit#(1) tx;
+  method Bit#(16) divisor;
+  method Vector#(cores, Bool) receiveInterrupt;
+endinterface
+
+module mkRVUart_RV32#(Bit#(16) divisor)(RVUart#(1))
+    provisos (NumAlias#(internalAddrSize, 8));
+  // Layout of memory interface (mostly derived by SiFive Freedom E300)
+  // Address   Name     Description
+  // 0x0000    txData   Transmit data register
+  // 0x0004    rxData   Receive data register
+  // 0x0008    txCtrl   Transmit control register
+  // 0x000C    rxCtrl   Receive control register
+  // 0x0010    div      Baud rate divisor
+  Bit#(internalAddrSize) txDataAddr = 'h0;
+  Bit#(internalAddrSize) rxDataAddr = 'h4;
+  Bit#(internalAddrSize) txCtrlAddr = 'h8;
+  Bit#(internalAddrSize) rxCtrlAddr = 'hC;
+  Bit#(internalAddrSize) divAddr    = 'h10;
+
+  // TODO: Make a circular ring buffer for that we can then pass to connectal for
+  //       memory mapped interface
+  // ie. txDataRing 0x000 -> index_1[31:24], index_2[23:16], index_3[15:8], index_4[7:0]
+  //     txTailPtr  0x004 -> ptr to whats next in the ring buffer
+  Ehr#(2, Maybe#(Tuple2#(Bool, Bit#(internalAddrSize)))) pendingReq <- mkEhr(tagged Invalid);
+
+  // BSV specific UART configuration
+  // BSV UART implementation does a few things
+  // its interface:
+  // interface UART#(numeric type depth);
+  //   interface RS232          rs232;
+  //     method sout = rXmitDataOut;
+  //     method sin  = rRecvData._write;
+  //   interface Get#(Bit#(8))  tx;
+  //     method ActionValue#(Bit#(8)) get;
+  //   interface Put#(Bit#(8))  rx;
+  //     method Action put(x)
+  // Get and Put have FiFos attached to the ends of them
+  Reg#(Bit#(16)) divReg <- mkReg(divisor);
+  UART#(16) uart <- mkUART(8, NONE, STOP_1, divReg);
+  // Data registers
+  Reg#(Maybe#(Bit#(8))) txDataReg <- mkReg(tagged Invalid);
+  Reg#(Maybe#(Bit#(8))) rxDataReg <- mkReg(tagged Invalid);
+  Reg#(Bit#(32)) txCtrlReg <- mkReg(0);
+  Reg#(Bit#(32)) rxCtrlReg <- mkReg(0);
```

Comment:
`txCtrlReg` and `rxCtrlReg` are not really used at the moment, but they may be useful in the future to enable tx or rx interrupts. For now if you would like to save registers without removing them completely, you could make them read only registers by replacing `mkReg` with `mkReadOnlyReg` from the `RegUtil` package provided by recycle-bsv-lib (https://github.com/csail-csg/recycle-bsv-lib/blob/master/src/bsv/RegUtil.bsv)